File: LeetCode/23.Merge_k_Sorted_Lists/23.Merge_k_Sorted_Lists.py
#!/usr/bin/env python3


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# heap
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        # min heap store element (val, ListNode)

        heap = []
        dummy = ListNode(-1)
        node = dummy

        for i, lst in enumerate(lists):
            if lst:
                heapq.heappush(heap, (lst.val, i, lst))

        if not heap:
            return None

        while heap:
            (val, i, list_node) = heapq.heappop(heap)

            node.next = list_node
            node = node.next

            if list_node.next:
                nxt = list_node.next
                heapq.heappush(heap, (nxt.val, i, nxt))

        return dummy.next


# divide and conquer
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        return self._merge_lists(lists, 0, len(lists) - 1)

    def _merge_lists(self, lists, left, right):
        if left == right:
            return lists[left]
        # divide
        mid = (left + right) // 2
        left_node = self._merge_lists(lists, left, mid)
        right_node = self._merge_lists(lists, mid + 1, right)
        # conquer
        return self._merge_node(left_node, right_node)

    def _merge_node(self, p, q):
        dummy = ListNode(-1)
        node = dummy

        while p and q:
            if p.val < q.val:
                node.next = p
                p = p.next
            else:
                node.next = q
                q = q.next
            node = node.next

        node.next = p if p else q
        return dummy.next

File: LeetCode/23.Merge_k_Sorted_Lists/23.Merge_k_Sorted_Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        int n = lists.length - 1;
        if (n < 0) {
            return null;
        }
        return mergeLists(lists, 0, n / 2, n);
    }

    private ListNode mergeLists(ListNode[] lists, int s, int m, int e) {
        // end condition
        if (s == e) {
            return lists[s];
        }

        // divide
        int l = (s + m) / 2;
        ListNode left = mergeLists(lists, s, l, m);
        int r = (m + 1 + e) / 2;
        ListNode right = mergeLists(lists, m + 1, r, e);

        // conquer
        ListNode dummy = new ListNode();
        ListNode node = dummy;
        ListNode lp = left;
        ListNode rp = right;

        while (lp != null && rp != null) {
            if (lp.val < rp.val) {
                node.next = lp;
                lp = lp.next;
            } else {
                node.next = rp;
                rp = rp.next;
            }
            node = node.next;
        }

        if (lp != null) {
            node.next = lp;
        }

        if (rp != null) {
            node.next = rp;
        }

        return dummy.next;
    }
}